File: testdata/python-sample/service.py
# service.py — order service for the Python sample
from typing import List, Optional
from models import (
    User, Product, Order, OrderStatus,
    Repository, Logger,
    calculate_total, format_order_summary,
)


class OrderService:
    """Handles order business logic."""

    def __init__(self, repo: Repository, logger: Logger) -> None:
        self.repo = repo
        self.logger = logger

    def place_order(self, user_id: int, product_ids: List[int]) -> Order:
        """Create a new order for the given user and products."""
        user = self._find_user(user_id)
        if not user.is_active():
            raise ValueError(f"User {user_id} is not active")

        products = self._fetch_products(product_ids)
        self._check_stock(products)

        order = self._build_order(user, products)
        validate_order(order)

        self.repo.save_order(order)
        self.logger.info("order placed successfully")
        return order

    def cancel_order(self, order_id: int) -> None:
        """Cancel an existing order."""
        orders = self.repo.list_orders(0)
        target: Optional[Order] = None
        for o in orders:
            if o.id == order_id:
                target = o
                break

        if target is None:
            raise ValueError(f"Order {order_id} not found")

        if not target.is_cancellable():
            raise ValueError(f"Order {order_id} cannot be cancelled in status {target.status.value}")

        self.repo.update_order_status(order_id, OrderStatus.CANCELLED)
        self.logger.info("order cancelled")

    def get_order_summary(self, order_id: int) -> str:
        """Return a formatted summary of an order."""
        orders = self.repo.list_orders(0)
        for o in orders:
            if o.id == order_id:
                return format_order_summary(o)
        raise ValueError(f"Order {order_id} not found")

    def _find_user(self, user_id: int) -> User:
        user = self.repo.find_user(user_id)
        if user is None:
            raise ValueError(f"User {user_id} not found")
        return user

    def _fetch_products(self, ids: List[int]) -> List[Product]:
        products = []
        for pid in ids:
            product = self.repo.find_product(pid)
            if product is None:
                raise ValueError(f"Product {pid} not found")
            products.append(product)
        return products

    def _check_stock(self, products: List[Product]) -> None:
        for p in products:
            if not p.is_available():
                raise ValueError(f"Product {p.id} is out of stock")

    def _build_order(self, user: User, products: List[Product]) -> Order:
        total = calculate_total(products)
        return Order(user_id=user.id, products=products, total=total)


def validate_order(order: Order) -> None:
    """Validate an order before processing."""
    if order is None:
        raise ValueError("order is None")
    if order.user_id <= 0:
        raise ValueError(f"invalid user_id: {order.user_id}")
    if not order.products:
        raise ValueError("order has no products")
    if order.total < 0:
        raise ValueError("order total cannot be negative")


class InMemoryRepository(Repository):
    """In-memory implementation of Repository for testing."""

    def __init__(self) -> None:
        self._users: dict = {}
        self._products: dict = {}
        self._orders: dict = {}
        self._next_id: int = 1

    def find_user(self, user_id: int) -> Optional[User]:
        return self._users.get(user_id)

    def find_product(self, product_id: int) -> Optional[Product]:
        return self._products.get(product_id)

    def save_order(self, order: Order) -> None:
        if order.id == 0:
            order.id = self._next_id
            self._next_id += 1
        self._orders[order.id] = order

    def list_orders(self, user_id: int) -> List[Order]:
        all_orders = list(self._orders.values())
        if user_id == 0:
            return all_orders
        return [o for o in all_orders if o.user_id == user_id]

    def update_order_status(self, order_id: int, status: OrderStatus) -> None:
        if order_id in self._orders:
            self._orders[order_id].status = status

    def add_user(self, user: User) -> None:
        self._users[user.id] = user

    def add_product(self, product: Product) -> None:
        self._products[product.id] = product

File: testdata/rust-sample/service.rs
// service.rs — Rust sample for code-graph test fixtures.
// Provides functions, impl blocks with trait implementations, and function calls.
// Must produce at least 8 symbols and 8 edges.

use std::fmt;

/// OrderService handles order business logic.
pub struct OrderService<R, L, N> {
    repo: R,
    logger: L,
    notifier: N,
}

impl<R, L, N> OrderService<R, L, N>
where
    R: Repository,
    L: Logger,
    N: Notifier,
{
    /// Create a new OrderService.
    pub fn new(repo: R, logger: L, notifier: N) -> Self {
        OrderService { repo, logger, notifier }
    }

    /// Place a new order for the given user and products.
    pub fn place_order(&mut self, user_id: u64, product_ids: &[u64]) -> Result<Order, String> {
        let user = self.repo.find_user(user_id).ok_or("user not found")?;
        if !user.is_active() {
            return Err(format!("user {} is not active", user_id));
        }
        let products = self.fetch_products(product_ids)?;
        check_stock(&products)?;
        let order = build_order(&user, products);
        validate_order(&order)?;
        self.repo.save_order(&order).map_err(|e| format!("save order: {}", e))?;
        self.logger.info("order placed successfully");
        let _ = self.notifier.notify_user(user_id, "Your order has been placed");
        Ok(order)
    }

    /// Cancel an existing order.
    pub fn cancel_order(&mut self, order_id: u64) -> Result<(), String> {
        let orders = self.repo.list_orders(0);
        let target = orders.into_iter().find(|o| o.id == order_id)
            .ok_or(format!("order {} not found", order_id))?;
        if !target.is_cancellable() {
            return Err(format!("order {} cannot be cancelled", order_id));
        }
        self.logger.info("order cancelled");
        let _ = self.notifier.notify_user(target.user_id, "Your order has been cancelled");
        Ok(())
    }

    /// Fetch products by their IDs.
    fn fetch_products(&self, ids: &[u64]) -> Result<Vec<Product>, String> {
        let mut products = Vec::new();
        for &id in ids {
            let p = self.repo.find_product(id).ok_or(format!("product {} not found", id))?;
            products.push(p);
        }
        Ok(products)
    }
}

/// Check that all products are in stock.
pub fn check_stock(products: &[Product]) -> Result<(), String> {
    for p in products {
        if !p.is_available() {
            return Err(format!("product {} is out of stock", p.id));
        }
    }
    Ok(())
}

/// Build an Order from a user and products.
pub fn build_order(user: &User, products: Vec<Product>) -> Order {
    let total = calculate_total(&products);
    Order {
        id: 0,
        user_id: user.id,
        products,
        total,
        status: OrderStatus::Pending,
    }
}

/// Sum the prices of all products.
pub fn calculate_total(products: &[Product]) -> f64 {
    products.iter().map(|p| p.price).sum()
}

/// Validate an order before processing.
pub fn validate_order(order: &Order) -> Result<(), String> {
    if order.user_id == 0 {
        return Err("invalid user ID".to_string());
    }
    if order.products.is_empty() {
        return Err("order has no products".to_string());
    }
    if order.total < 0.0 {
        return Err("order total cannot be negative".to_string());
    }
    Ok(())
}

/// Format a human-readable summary of an order.
pub fn format_order_summary(order: &Order) -> String {
    format!(
        "Order #{}: {} items, total ${:.2}",
        order.id,
        order.item_count(),
        order.total
    )
}

// ---- Minimal trait/type stubs so this file compiles standalone in tests ----

pub struct User {
    pub id: u64,
    pub name: String,
    pub status: UserStatus,
}

impl User {
    pub fn is_active(&self) -> bool {
        self.status == UserStatus::Active
    }
}

pub struct Product {
    pub id: u64,
    pub price: f64,
    pub stock: u32,
}

impl Product {
    pub fn is_available(&self) -> bool {
        self.stock > 0
    }
}

pub struct Order {
    pub id: u64,
    pub user_id: u64,
    pub products: Vec<Product>,
    pub total: f64,
    pub status: OrderStatus,
}

impl Order {
    pub fn is_cancellable(&self) -> bool {
        self.status == OrderStatus::Pending || self.status == OrderStatus::Confirmed
    }

    pub fn item_count(&self) -> usize {
        self.products.len()
    }
}

#[derive(PartialEq)]
pub enum UserStatus {
    Active,
    Inactive,
}

#[derive(PartialEq)]
pub enum OrderStatus {
    Pending,
    Confirmed,
    Cancelled,
}

pub trait Repository {
    fn find_user(&self, id: u64) -> Option<User>;
    fn find_product(&self, id: u64) -> Option<Product>;
    fn save_order(&mut self, order: &Order) -> Result<(), String>;
    fn list_orders(&self, user_id: u64) -> Vec<Order>;
}

pub trait Logger {
    fn info(&self, msg: &str);
}

pub trait Notifier {
    fn notify_user(&self, user_id: u64, message: &str) -> Result<(), String>;
}
